File: .claude/hooks/file-changed-run-tests.mjs
#!/usr/bin/env node
// @hookstack file-changed-run-tests
import { execSync } from "node:child_process";
// @hookstack file-changed-run-tests
// Relance les tests impactés quand un fichier source change (FileChanged)
// Si vitest est disponible : `vitest related --run <fichier>` ne rejoue que les
// tests qui importent le fichier modifié (latence et tokens minimaux).
// Sinon, repli sur le script `test` du gestionnaire de paquets du projet.
// CI=true force le mode non-watch (évite un hook qui pend jusqu'au timeout).
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

function defaultExec(cmd) {
	return execSync(cmd, {
		timeout: 90_000,
		env: { ...process.env, CI: "true" },
	});
}

// Détecte le gestionnaire de paquets depuis le lockfile (cohérent avec enforce-package-managers).
export function detectManager({
	exists = existsSync,
	projectDir = process.cwd(),
} = {}) {
	if (exists(join(projectDir, "pnpm-lock.yaml"))) return "pnpm";
	if (
		exists(join(projectDir, "bun.lockb")) ||
		exists(join(projectDir, "bun.lock"))
	)
		return "bun";
	if (exists(join(projectDir, "yarn.lock"))) return "yarn";
	return "npm";
}

export function run(
	input,
	{
		exec = defaultExec,
		exists = existsSync,
		projectDir = process.env.CLAUDE_PROJECT_DIR ?? process.cwd(),
	} = {},
) {
	if (input.event === "unlink") return null;

	const filePath = input.file_path ?? "";
	const hasVitest = exists(join(projectDir, "node_modules", ".bin", "vitest"));
	const cmd =
		hasVitest && filePath
			? `npx --no-install vitest related --run "${filePath}" 2>&1`
			: `${detectManager({ exists, projectDir })} test --if-present 2>&1`;

	try {
		const out = exec(cmd).toString();
		return {
			hookSpecificOutput: {
				hookEventName: "FileChanged",
				additionalContext: `Tests passed after ${filePath} changed.\n${out.slice(-500)}`,
			},
		};
	} catch (e) {
		const out = (e.stdout ?? e.stderr ?? e.message).toString().slice(0, 1000);
		return {
			hookSpecificOutput: {
				hookEventName: "FileChanged",
				additionalContext: `Tests FAILED after ${filePath} changed:\n${out}`,
			},
		};
	}
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
	const input = JSON.parse(readFileSync(0, "utf8"));
	const result = run(input);
	if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/file-changed-run-tests.py
#!/usr/bin/env python3
# @hookstack file-changed-run-tests
"""Re-runs the affected tests when a source file changes (FileChanged)."""
import json
import os
import subprocess
import sys


def _exec(cmd):
    env = dict(os.environ)
    env["CI"] = "true"
    return subprocess.run(
        cmd, shell=True, capture_output=True, text=True, timeout=90, check=True, env=env
    ).stdout


def _exists(path):
    return os.path.exists(path)


def run(input_data, *, exec_cmd=None, exists=None, project_dir=None):
    if exec_cmd is None:
        exec_cmd = _exec
    if exists is None:
        exists = _exists
    if project_dir is None:
        project_dir = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())

    if input_data.get("event") == "unlink":
        return None

    file_path = input_data.get("file_path") or ""

    # Projet Python : pytest via uv (préféré) ou python3 en repli.
    has_pyproject = exists(os.path.join(project_dir, "pyproject.toml"))
    if has_pyproject:
        cmd = "uv run pytest -q 2>&1"
    else:
        cmd = "python3 -m pytest -q 2>&1"

    try:
        out = exec_cmd(cmd)
        return {
            "hookSpecificOutput": {
                "hookEventName": "FileChanged",
                "additionalContext": f"Tests passed after {file_path} changed.\n{out[-500:]}",
            }
        }
    except Exception as e:
        out = (getattr(e, "stdout", None) or getattr(e, "stderr", None) or str(e))[:1000]
        return {
            "hookSpecificOutput": {
                "hookEventName": "FileChanged",
                "additionalContext": f"Tests FAILED after {file_path} changed:\n{out}",
            }
        }


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
